File: android/app/src/main/java/com/agentcontrolcenter/app/runtime/task/TaskManager.kt
package com.agentcontrolcenter.app.runtime.task

import com.agentcontrolcenter.app.core.database.dao.TaskDao
import com.agentcontrolcenter.app.core.database.entity.TaskEntity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 任务管理器 — 管理下发给 Agent 的异步任务。
 *
 * 与 Chat 的区别：
 * - Chat 是同步的请求-响应流
 * - Task 是异步的、可追踪的、可调度的执行单元
 *
 * 例如：
 * - "让 OpenCode 重构这个文件" → Task
 * - "让 Ollama 生成图片" → Task
 * - "让 OpenManus 执行工作流" → Task
 *
 * Phase 4.2: 通过 [taskDao] 持久化任务到 Room。
 * App 重启后任务历史不丢失。内存中的 [_tasks] StateFlow 仍保留用于
 * 即时 UI 反馈，但真实数据源是数据库。
 */
@Singleton
class TaskManager @Inject constructor(
    private val taskDao: TaskDao
) {

    data class AgentTask(
        val id: String,
        val agentId: String,
        val sessionId: String?,
        val type: TaskType,
        val input: String,
        val status: TaskStatus = TaskStatus.Pending,
        val result: String? = null,
        val createdAt: Long = System.currentTimeMillis(),
        val completedAt: Long? = null,
        val error: String? = null
    )

    enum class TaskType { CHAT, CODE, WORKFLOW, TOOL_CALL, FILE_OPERATION }
    enum class TaskStatus { Pending, Running, Completed, Failed, Cancelled }

    private val supervisorScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _tasks = MutableStateFlow<List<AgentTask>>(emptyList())
    val tasks: StateFlow<List<AgentTask>> = _tasks.asStateFlow()

    init {
        // 订阅数据库变更，同步到内存 StateFlow
        supervisorScope.launch {
            taskDao.getAllTasks().collect { entities ->
                _tasks.value = entities.map { it.toModel() }
            }
        }
    }

    fun submitTask(agentId: String, type: TaskType, input: String, sessionId: String? = null): AgentTask {
        val task = AgentTask(
            id = "task_${UUID.randomUUID()}",
            agentId = agentId,
            sessionId = sessionId,
            type = type,
            input = input
        )
        // 持久化到数据库
        supervisorScope.launch {
            taskDao.upsertTask(task.toEntity())
        }
        // 同时更新内存（数据库 Flow 回调也会更新，但这里先更新保证即时反馈）
        _tasks.update { it + task }
        return task
    }

    fun updateTaskStatus(taskId: String, status: TaskStatus, result: String? = null, error: String? = null) {
        val completedAt = if (status == TaskStatus.Completed || status == TaskStatus.Failed || status == TaskStatus.Cancelled) {
            System.currentTimeMillis()
        } else null

        // 持久化到数据库
        supervisorScope.launch {
            taskDao.updateTaskStatus(
                taskId = taskId,
                status = status.name,
                result = result,
                error = error,
                completedAt = completedAt
            )
        }
        // 同时更新内存
        _tasks.update { list ->
            list.map { task ->
                if (task.id == taskId) {
                    task.copy(
                        status = status,
                        result = result ?: task.result,
                        error = error ?: task.error,
                        completedAt = completedAt
                    )
                } else task
            }
        }
    }

    fun getTasksForAgent(agentId: String): List<AgentTask> {
        return _tasks.value.filter { it.agentId == agentId }
    }

    fun getActiveTasks(): List<AgentTask> {
        return _tasks.value.filter { it.status == TaskStatus.Pending || it.status == TaskStatus.Running }
    }

    fun cancelTask(taskId: String) {
        updateTaskStatus(taskId, TaskStatus.Cancelled)
    }

    /** 删除任务（从数据库和内存中移除）。 */
    fun deleteTask(taskId: String) {
        supervisorScope.launch {
            taskDao.deleteTask(taskId)
        }
        _tasks.update { list -> list.filter { it.id != taskId } }
    }

    // ── Entity ↔ Model 转换 ──

    private fun AgentTask.toEntity() = TaskEntity(
        id = id,
        agentId = agentId,
        sessionId = sessionId,
        type = type.name,
        input = input,
        status = status.name,
        result = result,
        createdAt = createdAt,
        completedAt = completedAt,
        error = error
    )

    private fun TaskEntity.toModel() = AgentTask(
        id = id,
        agentId = agentId,
        sessionId = sessionId,
        type = try { TaskType.valueOf(type) } catch (_: Exception) { TaskType.CHAT },
        input = input,
        status = try { TaskStatus.valueOf(status) } catch (_: Exception) { TaskStatus.Pending },
        result = result,
        createdAt = createdAt,
        completedAt = completedAt,
        error = error
    )
}

File: ios/AgentControlCenter/Runtime/TaskManager.swift
import Foundation
import Observation

// MARK: - TaskManager
// 对应 Android TaskManager (Phase 4.2)

/// 任务管理器 — 管理下发给 Agent 的异步任务。
///
/// 与 Chat 的区别：
/// - Chat 是同步的请求-响应流
/// - Task 是异步的、可追踪的、可调度的执行单元
///
/// 例如：
/// - "让 OpenCode 重构这个文件" → Task
/// - "让 OpenAI 生成图片" → Task
/// - "让 Hermes 执行工作流" → Task
///
/// Phase 4.2 说明：
/// - Android 版通过 Room TaskDao 持久化任务到数据库
/// - iOS 版当前使用内存存储，后续可通过 SwiftData / CoreData 补齐持久化
/// - 数据模型复用 Models/Task.swift 中的 AgentTask / TaskType / TaskStatus
///
/// `@MainActor` 隔离保证 `tasks` 等响应式状态的读写均在主线程进行，
/// 避免 SwiftUI 视图读取时发生数据竞争。
@MainActor
@Observable
final class TaskManager {

    /// 所有任务列表
    private(set) var tasks: [AgentTask] = []

    // MARK: - 任务提交

    /// 提交新任务
    /// - Parameters:
    ///   - agentId: 目标 Agent ID
    ///   - type: 任务类型
    ///   - input: 任务输入内容
    ///   - sessionId: 关联的会话 ID（可选）
    /// - Returns: 新创建的任务
    @discardableResult
    func submitTask(
        agentId: String,
        type: TaskType,
        input: String,
        sessionId: String? = nil
    ) -> AgentTask {
        let task = AgentTask(
            id: "task_\(UUID().uuidString)",
            agentId: agentId,
            sessionId: sessionId,
            type: type,
            input: input
        )
        tasks.append(task)
        return task
    }

    // MARK: - 状态更新

    /// 更新任务状态
    /// - Parameters:
    ///   - taskId: 任务 ID
    ///   - status: 新状态
    ///   - result: 任务结果（可选，终态时填充）
    ///   - error: 错误信息（可选，失败时填充）
    func updateStatus(
        taskId: String,
        status: TaskStatus,
        result: String? = nil,
        error: String? = nil
    ) {
        guard let index = tasks.firstIndex(where: { $0.id == taskId }) else { return }
        var task = tasks[index]
        task.status = status
        // 仅在提供新值时更新 result 和 error，保留已有值
        if let result = result { task.result = result }
        if let error = error { task.error = error }
        // 终态（完成/失败/取消）时记录完成时间
        if status.isTerminal {
            task.completedAt = Int64(Date().timeIntervalSince1970 * 1000)
        }
        tasks[index] = task
    }

    // MARK: - 查询

    /// 获取指定 Agent 的所有任务
    /// - Parameter agentId: Agent ID
    /// - Returns: 该 Agent 的任务列表
    func getTasksForAgent(_ agentId: String) -> [AgentTask] {
        tasks.filter { $0.agentId == agentId }
    }

    /// 当前活跃任务（待处理或运行中）
    var activeTasks: [AgentTask] {
        tasks.filter { $0.status == .pending || $0.status == .running }
    }

    // MARK: - 任务控制

    /// 取消任务
    /// - Parameter taskId: 任务 ID
    func cancelTask(_ taskId: String) {
        updateStatus(taskId: taskId, status: .cancelled)
    }

    /// 删除任务
    /// - Parameter taskId: 任务 ID
    func deleteTask(_ taskId: String) {
        tasks.removeAll { $0.id == taskId }
    }
}
